clustered/client/src/main/java/org/ehcache/clustered/client/internal/EhcacheClientEntity.java

```diff
@@ -90,7 +90,7 @@
   private final LifeCycleMessageFactory messageFactory;
   private final Map<Class<? extends EhcacheEntityResponse>, List<ResponseListener<? extends EhcacheEntityResponse>>> responseListeners = new ConcurrentHashMap<Class<? extends EhcacheEntityResponse>, List<ResponseListener<? extends EhcacheEntityResponse>>>();
   private final List<DisconnectionListener> disconnectionListeners = new CopyOnWriteArrayList<DisconnectionListener>();
-  private final List<ReconnectListener> reconnectListeners = new ArrayList<ReconnectListener>();
+  private final List<ReconnectListener> reconnectListeners = new CopyOnWriteArrayList<ReconnectListener>();
```

Comment:
Nope i.e. not the reason. Let me explain it again, it is to keep store creation on hold while reconnect data is being formed. CopyWrite does provides immutable data, so its not about exclusive access. The Store completion at client side creates store at server side, and then registers the listeners(kind of two step process).So reconnect can interleave during cache creatio itself. With CopyOnWrite, listener registration will complete in middle of reconnect data being formed, and cache can theoretically start doing operations which is absent in reconnect data.Though with eventual invalidation accounting on the server side, this may be rare with strong caches, but for now I would suggest to leave it with sync so that its easier to reason about it. IMHO, disconnection listener registration/deregistraion should also be revisited w.r.t. to     race with didDisconnectUnexpectedly event being fired.